sycl/include/CL/sycl/intel/function_pointer.hpp

```diff
@@ -0,0 +1,99 @@
+//==----------- function_pointer.hpp --- SYCL Function pointers ------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/device.hpp>
+#include <CL/sycl/program.hpp>
+#include <CL/sycl/stl.hpp>
+
+#include <type_traits>
+
+namespace cl {
+namespace sycl {
+namespace intel {
+
+// This is a preview extension implementation, intended to provide early access
+// to a feature for review and community feedback.
+//
+// Because the interfaces defined by this header file are not final and are
+// subject to change they are not intended to be used by shipping software
+// products. If you are interested in using this feature in your software
+// product, please let us know!
+
+using device_func_ptr_holder_t = cl::sycl::cl_ulong;
+
+/// \brief this function performs a cast from device_func_ptr_holder_t type
+/// to the provided function pointer type.
+template <
+    class FuncType,
+    typename FuncPtrType = typename std::add_pointer<FuncType>::type,
+    typename std::enable_if<std::is_function<FuncType>::value, int>::type = 0>
+inline FuncPtrType to_device_func_ptr(device_func_ptr_holder_t FptrHolder) {
+  return reinterpret_cast<FuncPtrType>(FptrHolder);
+}
+
+template <class FuncType>
+using enable_if_is_function_pointer_t = typename std::enable_if<
+    std::is_pointer<FuncType>::value &&
+        std::is_function<typename std::remove_pointer<FuncType>::type>::value,
+    int>::type;
+
+/// \brief this function can be used only on host side to obtain device function
+/// pointer for the specified function.
+///
+/// \param F - pointer to function to make it work for SYCL Host device
+/// \param FuncName - name of the function. Please note that by default names of
+/// functions are mangled since SYCL is a C++. To avoid the need ot specifying
+/// mangled name here, use `extern "C"`
+/// \param P - sycl::program object which will be used to extract device
+/// function pointer
+/// \param D - sycl::device object which will be used to extract device
+/// function pointer
+///
+/// \returns device_func_ptr_holder_t object which can be used inside a device
+/// code. This object must be converted back to a function pointer using
+/// `to_device_func_ptr` prior to actual usage.
+///
+/// Returned value is valid only within device code which was compiled for the
+/// specified program and device. Returned value invalidates whenever program
+/// is released or re-built
+template <class FuncType, enable_if_is_function_pointer_t<FuncType> = 0>
+device_func_ptr_holder_t get_device_func_ptr(FuncType F, const char *FuncName,
+                                             program &P, device &D) {
```

Comment:
Certain OpenCL implementations just-in-time (JIT) compile kernels on the `clEnqueueNDRangeKernel` call when all scheduling and argument information is available. This makes device functions pointers of non-enqueued kernels of limited use/unusable.

What does having access to a device function pointer enable that isn't achievable with the existing kernel object?

Reply:
> > I can refactor this into a new PI interface if you prefer, but I think that it would be good to leave this
> 
> Yes, please add piDeviceGetFunctionPointer or similar.
> 
> We should be able to add new PI interfaces as new features are added, so I think we don't need this generic extension query. @bader , @romanovvlad what do you think?

So your suggestion is to hide any extension loading behind always-available PI functions?
If they're not available, the function should return an error or something?

Also, this PR could use a test.

Reply:
> Certain OpenCL implementations just-in-time (JIT) compile kernels on the `clEnqueueNDRangeKernel` call when all scheduling and argument information is available. This makes device functions pointers of non-enqueued kernels of limited use/unusable.

Just a few ideas how to overcome this limitations:
- perform JIT compilation of the function which address was requested and make sure that it is available from kernels that will be JIT-ed later. On CPUs it is probably not a problem at all if all JIT-ed objects live within the same process
- since the return value is a some kind of "holder" object, implementation may return some "handle" value and resolve it at runtime (by JIT-ing once more) once it was reinterpreted as function pointer and called.
- perform full JIT of everything if this extension is in use.
- combination of above?

> What does having access to a device function pointer enable that isn't achievable with the existing kernel object?

For example, you can create some kind of dispatch table which is filled on host side and then used by device as "callback" functions, see `pass-fp-through-buffer.cpp` test in this PR